From f9505ea962919a4c4969b3dce1ad46eca75a10c3 Mon Sep 17 00:00:00 2001
From: Denys Vlasenko <dvlasenk@redhat.com>
Date: Thu, 4 Aug 2011 12:07:15 +0200
Subject: [PATCH 28/52] add 'report' compat tool. Closes #315, bz#725660.

Tested with RHTsupport. Bugzilla likely doesn't work, pending
addition od "reporter-bugzilla -t TICKET" support.

Signed-off-by: Denys Vlasenko <dvlasenk@redhat.com>
---
 src/include/internal_libreport.h |    2 +-
 src/plugins/Makefile.am          |   23 +++++++++++-
 src/plugins/report.c             |   73 ++++++++++++++++++++++++++++++++++++++
 4 files changed, 109 insertions(+), 2 deletions(-)
 create mode 100644 src/plugins/report.c

--- a/src/include/internal_libreport.h
+++ b/src/include/internal_libreport.h
@@ -703,7 +703,7 @@ unsigned parse_opts(int argc, char **argv, const struct options *opt,
                 const char *usage);

 #define show_usage_and_die libreport_show_usage_and_die
-void show_usage_and_die(const char *usage, const struct options *opt);
+void show_usage_and_die(const char *usage, const struct options *opt) NORETURN;


 #ifdef __cplusplus
diff --git a/src/plugins/Makefile.am b/src/plugins/Makefile.am
index 02e0647..3883a30 100644
--- a/src/plugins/Makefile.am
+++ b/src/plugins/Makefile.am
@@ -8,7 +8,8 @@ bin_PROGRAMS = \
     reporter-kerneloops \
     reporter-upload \
     reporter-mailx \
-    reporter-print
+    reporter-print \
+    report

 pluginsconfdir = $(PLUGINS_CONF_DIR)

@@ -206,4 +207,24 @@ reporter_print_CPPFLAGS = \
 reporter_print_LDADD = \
     ../lib/libreport.la

+report_SOURCES = \
+    report.c
+report_CPPFLAGS = \
+    -I$(srcdir)/../include \
+    -I$(srcdir)/../lib \
+    -DBIN_DIR=\"$(bindir)\" \
+    -DVAR_RUN=\"$(VAR_RUN)\" \
+    -DCONF_DIR=\"$(CONF_DIR)\" \
+    -DLOCALSTATEDIR='"$(localstatedir)"' \
+    -DDEBUG_DUMPS_DIR=\"$(DEBUG_DUMPS_DIR)\" \
+    -DDEBUG_INFO_DIR=\"$(DEBUG_INFO_DIR)\" \
+    -DPLUGINS_LIB_DIR=\"$(PLUGINS_LIB_DIR)\" \
+    -DPLUGINS_CONF_DIR=\"$(REPORT_PLUGINS_CONF_DIR)\" \
+    $(GLIB_CFLAGS) \
+    $(LIBREPORT_CFLAGS) \
+    -D_GNU_SOURCE \
+    -Wall -Wwrite-strings -Werror
+report_LDADD = \
+    ../lib/libreport.la
+
 DEFS = -DLOCALEDIR=\"$(localedir)\" @DEFS@
diff --git a/src/plugins/report.c b/src/plugins/report.c
new file mode 100644
index 0000000..2ba5448
--- /dev/null
+++ b/src/plugins/report.c
@@ -0,0 +1,73 @@
+/*
+    Copyright (C) 2011  ABRT team
+    Copyright (C) 2011  RedHat Inc
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+#include "internal_libreport.h"
+
+int main(int argc, char **argv)
+{
+    abrt_init(argv);
+
+    const char *target = NULL;
+    const char *ticket = NULL;
+
+    /* Can't keep these strings/structs static: _() doesn't support that */
+    const char *program_usage_string = _(
+        "\b [-v] --target TARGET --ticket ID FILE...\n"
+        "\n"
+        "Uploads FILEs to specified ticket on TARGET.\n"
+        "\n"
+        "This tool is provided to ease transition of users of report package\n"
+        "to libreport. Recognized TARGETs are 'strata' and 'bugzilla',\n"
+        "first one invokes upload to RHTSupport and second - to Bugzilla."
+    );
+    enum {
+        OPT_v = 1 << 0,
+        OPT_T = 1 << 1,
+        OPT_t = 1 << 2,
+    };
+    /* Keep enum above and order of options below in sync! */
+    struct options program_options[] = {
+        OPT__VERBOSE(&g_verbose),
+        OPT_STRING('T', "target", &target, "TARGET", _("'strata' or 'bugzilla'")),
+        OPT_STRING('t', "ticket", &ticket, "ID"    , _("Ticket/case ID")),
+        OPT_END()
+    };
+    /*unsigned opts =*/ parse_opts(argc, argv, program_options, program_usage_string);
+
+    export_abrt_envvars(0);
+
+    argv += optind;
+    if (!*argv || !target || !ticket)
+        show_usage_and_die(program_usage_string, program_options);
+
+    const char *tool_name;
+    if (strcmp(target, "strata") == 0)
+        tool_name = "reporter-rhtsupport";
+    else
+    if (strcmp(target, "bugzilla") == 0)
+        tool_name = "reporter-bugzilla";
+    else
+        show_usage_and_die(program_usage_string, program_options);
+
+    argv -= 2;
+    argv[0] = (char*) tool_name;
+    argv[1] = xasprintf("-t%s", ticket);
+
+    execvp(argv[0], argv);
+    perror_msg_and_die("Can't execute '%s'", argv[0]);
+}
--
1.7.6

